File: leetcode/medium/1865. Finding Pairs With a Certain Sum/solution.js
class FindSumPairs {
    /**
    * @param {number[]} nums1
    * @param {number[]} nums2
    */
    constructor(nums1, nums2) {
        this.nums1 = nums1;
        this.nums2 = nums2;
        this.m1 = this.mapping(nums1);
        this.m2 = this.mapping(nums2);
    }

    /**
     * @param {number[]} arr 
     * @returns {map} map
     */
    mapping(arr) {
        const map = new Map();
        arr.forEach(val => map.set(val, map.get(val) + 1 || 1));
        return map;
    }

    /**
     * @param {number} index 
     * @param {number} val 
     * @return {void} void
     */
    add(index, val) {
        const prev = this.nums2[index];
        const current = this.nums2[index] + val;
        this.nums2[index] = current;

        const freq = this.m2.get(prev);
        freq > 1 ? this.m2.set(prev, freq - 1) : this.m2.delete(prev);

        this.m2.set(current, this.m2.get(current) + 1 || 1);
    };

    /**
     * @param {number} tot 
     * @returns {number} number
     */
    count(tot) {
        const keys = Array.from(this.m1.keys());
        let res = 0;
        for (const key of keys) {
            const diff = tot - key;
            if (this.m2.has(diff)) {
                res += this.m1.get(key) * this.m2.get(diff);
            }
        }
        return res;
    }
}

File: leetcode/medium/1865. Finding Pairs With a Certain Sum/solution.cc
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class FindSumPairs {
public:
    unordered_map<int, int> umap;
    vector<int> n1, n2;

    FindSumPairs(vector<int>& nums1, vector<int>& nums2) {
        swap(n1, nums1);
        swap(n2, nums2);

        sort(n1.begin(), n1.end());
        for (auto n: n2) {
            umap[n] += 1;
        }
    }
    
    void add(int index, int val) {
        umap[n2[index]] -= 1;
        n2[index] += val;
        umap[n2[index]] += 1;
    }
    
    int count(int tot) {
        int res = 0;
        for (auto i = 0; i < n1.size() && n1[i] < tot; i += 1) {
            auto it = umap.find(tot - n1[i]);
            if (it != end(umap)) res += it->second;
        }
        return res;
    }
};

int main() {
    return 0;
}
/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs* obj = new FindSumPairs(nums1, nums2);
 * obj->add(index,val);
 * int param_2 = obj->count(tot);
 */